apps/drupal/src/Plugin/Preprocessor/PreprocessorManager.php

```diff
@@ -0,0 +1,500 @@
+    $types = array_slice($directory, 3);
+    foreach ($types as $type) {
+      $items = scandir(static::getPreprocessorsDirectory() . '/' . $type);
+      foreach ($items as $item) {
+        if (!is_dir($item)) {
+          $item = str_replace('.php', '', $item);
+          $preprocessors[$item] = $type;
+        }
+      }
+    }
+
+    return $preprocessors;
+  }
+
+  /**
+   * Get the preprocessor class map from State API or in memory store.
+   *
+   * @return array
+   *   Array of keys of preprocess name and values of full class path or empty.
+   */
+  public static function getPreprocessorsClassMap() {
+    $preprocessor_map = &drupal_static(__FUNCTION__);
+    if (is_null($preprocessor_map)) {
+      $preprocessor_map = \Drupal::state()->get(static::STATE_ID);
+      if ($preprocessor_map && is_array($preprocessor_map)) {
+        return $preprocessor_map;
+      }
+    }
+
+    return isset($preprocessor_map) ? $preprocessor_map : [];
+  }
+
+  /**
+   * Merge suggestions with a matching Preprocessor Class into Preprocessor Map.
+   *
+   * @param array $suggestions
+   *   The suggestions from Drupal.
+   *
+   * @param array $preprocessor_map
+   *   The cached state api map.
+   *
+   * @return array
+   *   Merged Map combination of original Map and Classes.
+   */
+  public static function mergedMaps(array $suggestions, array $preprocessor_map) {
+    $map = [];
+    // If the Processor Matches a Suggestion, Merge into Preprocessor Map.
+    if ($preprocessors = static::getPreprocessors()) {
+      foreach ($preprocessors as $preprocessor => $type) {
+        if (!is_dir($preprocessor)) {
+          $key = array_search(strtolower($preprocessor), $suggestions);
+          if ($key) {
+            $map[$key] = __NAMESPACE__ . '\\Preprocessors' . '\\' . $type . '\\' . $preprocessor;
+          }
+        }
+      }
+    }
+
+    return array_unique(array_merge($preprocessor_map, $map));
+  }
+
+  /**
+   * Make the preprocessor class map.
+   *
+   * @param array $preprocessor_map
+   *   The cached state api map.
+   *
+   * @param array $new_suggestions
+   *   The suggestions from Drupal.
+   *
+   * @return array
+   *   An array with keys of preprocess name and values of full class path.
+   */
+  public static function makePreprocessorsClassMap(array $preprocessor_map, array $new_suggestions = []) {
+    // Initialize Preprocessors Map with Base.
+    if (!array_key_exists('default', $preprocessor_map)) {
+      $map['default'] = __NAMESPACE__ . '\\Preprocessors\\Base';
+      drupal_static_reset();
+      \Drupal::state()->set(static::STATE_ID, $map);
+      $preprocessor_map = static::getPreprocessorsClassMap();
+    }
+
+    // If we have new suggestions, try to update the preprocessor map state.
+    if (!empty($new_suggestions)) {
+      $new_maps = static::mergedMaps($new_suggestions, $preprocessor_map);
+      if (!empty($new_maps) && $new_maps !== $preprocessor_map) {
```

Comment:
@jryanconklin Both arrays need to be sorted by **value** before checking for equality in PHP. This could be done in the getPreprocessorsClassMap() and mergedMaps().
Example:
`
$x = ['a' => 1, 'b' => 2];
$y = ['a' => 1, 'b' => 2];
$z = ['b' => 2, 'a' => 1];

$x === $y; // returns TRUE
$x === $z; // returns FALSE

`

Reply:
Updated! Added an asort which should ensure these are in the same state.